colibre/scripts/density_species_interp.py

```diff
@@ -0,0 +1,339 @@
+"""
+Makes a rho-T plot. Uses the swiftsimio library.
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import glob
+
+import swiftsimio as sw
+print(sw.__file__)
+from swiftsimio import load
+import h5py as h5
+
+from unyt import mh, cm
+from matplotlib.colors import LogNorm, ListedColormap, BoundaryNorm
+from matplotlib.animation import FuncAnimation
+from matplotlib.cm import get_cmap
+from scipy.interpolate import interpn
+
+# Set the limits of the figure.
+density_bounds = [1e-3, 1e3]  # in nh/cm^3
+temperature_bounds = [10 ** (0), 10 ** (9.5)]  # in K
+dustfracs_bounds = [-4, -1]  # In metal mass fraction
+min_dustfracs = 10 ** dustfracs_bounds[0]
+bins = 64
+elements = [
+    "Hydrogen",
+    "Helium",
+    "Carbon",
+    "Nitrogen",
+    "Oxygen",
+    "Neon",
+    "Magnesium",
+    "Silicon",
+    "Sulphur",
+    "Calcium",
+    "Iron",
+    "Oatoms",]
+
+
+def get_data(filename, tables, prefix_rho, prefix_T):
+    """
+    Grabs the data (T in Kelvin and density in mh / cm^3) and interpolates for species fractions..
+    """
+
+    data = load(filename)
+    number_density = (
+        getattr(data.gas, f"{prefix_rho}densities").to_physical() / mh
+    ).to(cm ** -3)
+    temperature = getattr(data.gas, f"{prefix_T}temperatures").to_physical().to("K")
+    masses = data.gas.masses.to_physical().to("Msun")
+    
+    # Abundance Mass Fractions 
+    X = data.gas.element_mass_fractions.hydrogen
+    Z = data.gas.metal_mass_fractions
+
+    # Species Mass Fractions
+    neutfrac = X * masses * data.gas.species_fractions.HI
+    molfrac = 2 * X * masses * data.gas.species_fractions.H2
+    ionfrac = X * masses * data.gas.species_fractions.HII
+
+    # Dust Mass Fractions
+    dfracs = np.zeros(data.gas.masses.shape)
+    for d in dir(data.gas.dust_mass_fractions):
+        if hasattr(getattr(data.gas.dust_mass_fractions, d), "units"):
+            dfracs += getattr(data.gas.dust_mass_fractions, d)
+            
+    if glob.glob(tables):
+        table_file = h5.File(tables, "r")
```

Comment:
Please use `with h5.File(tables, "r")` rather than explitcitly opening and closing the file. That ensures if it crashes, or an exception is raised, that the file is closed 'nicely'

Reply:
done